pinot-core/src/main/java/org/apache/pinot/core/data/manager/offline/DimensionTableDataManager.java

```diff
@@ -0,0 +1,127 @@
+ * which can be accessed via 'getInstanceByTableName' static method.
+ */
+public class DimensionTableDataManager extends OfflineTableDataManager {
+  // Store singletons per table in this map
+  private static final Map<String, DimensionTableDataManager> _instances = new ConcurrentHashMap<>();
+
+  public static DimensionTableDataManager createInstanceByTableName(String tableName) {
+    _instances.putIfAbsent(tableName, new DimensionTableDataManager());
+    return _instances.get(tableName);
+  }
+
+  public static DimensionTableDataManager getInstanceByTableName(String tableName) {
+    return _instances.get(tableName);
+  }
+
+  // DimensionTableDataManager Instance properties
+  //
+
+  // _lookupTable is a HashMap used for fetching records from a table given the primary key
+  private final Map<PrimaryKey, GenericRow> _lookupTable = new HashMap<>();
+  private final ReadWriteLock _rwl = new ReentrantReadWriteLock();
+  private final Lock _lookupTableReadLock = _rwl.readLock();
+  private final Lock _lookupTableWriteLock = _rwl.writeLock();
+  private List<String> _primaryKeyColumns;
+
+  @Override
+  protected void doInit() {
+    super.doInit();
+
+    Schema schema = ZKMetadataProvider.getTableSchema(_propertyStore, _tableNameWithType);
+    Preconditions.checkState(schema != null, "Failed to find schema for table: %s", _tableNameWithType);
+    _primaryKeyColumns = schema.getPrimaryKeyColumns();
+    Preconditions.checkState(!CollectionUtils.isEmpty(_primaryKeyColumns),
+        "Primary key columns must be configured for dimension tables");
+  }
+
+  @Override
+  public void addSegment(File indexDir, IndexLoadingConfig indexLoadingConfig)
+      throws Exception {
+    super.addSegment(indexDir, indexLoadingConfig);
+    try {
+      prepareLookupTable();
+      _logger.info("Successfully loaded lookup table for {}", getTableName());
+    } catch (Exception e) {
+      throw new RuntimeException(
+          String.format("Error loading lookup table: {}", getTableName()),e);
+    }
+  }
+
+  private void prepareLookupTable() throws Exception {
+    List<SegmentDataManager> segmentManagers = acquireAllSegments();
+    List<File> indexDirs = new ArrayList<>();
+
+    for (SegmentDataManager segmentManager: segmentManagers) {
+      IndexSegment indexSegment = segmentManager.getSegment();
+      indexDirs.add(indexSegment.getSegmentMetadata().getIndexDir());
+    }
+    MultiplePinotSegmentRecordReader reader = new MultiplePinotSegmentRecordReader(indexDirs);
+
+    _lookupTableWriteLock.lock();
+    try {
+      _lookupTable.clear();
+      while (reader.hasNext()) {
+        GenericRow row = reader.next();
+        _lookupTable.put(row.getPrimaryKey(_primaryKeyColumns), row);
+      }
+    } finally {
+      _lookupTableWriteLock.unlock();
+    }
+  }
+
+  public GenericRow lookupRowByPrimaryKey(PrimaryKey pk) {
+    _lookupTableReadLock.lock();
```

Comment:
Use `ReadWriteLock` (https://docs.oracle.com/javase/7/docs/api/java/util/concurrent/locks/ReadWriteLock.html) to allow multiple readers?

Reply:
Already using a `ReentrantReadWriteLock();` for this. Did you mean to suggest switching to a different implementation?

Reply:
oh, sorry. missed line 71. How often does the write happen? Do you plan for a periodical or on-demand load? That may determine whether we shall enable the fair mode.

Reply:
For now the only entry point to the write operation is `addSegment` hook (I will be adding `removeSegment` too). Since the dim table is modeled as an OFFLINE table, writes will be very rare compared to reads. I'm thinking a strict fairness shouldn't be necessary in this context but happy to hear your thoughts.

On a related note, I realized I need to guard the whole `prepareLookupTable` body (not just the HashMap update section) with the write lock in case of burst calls to `addSegment`.